src/models/URL.js

```diff
@@ -0,0 +1,400 @@
+                ]) : def
+            }))
+            .set('password', new Parameter({
+                key: 'password',
+                type: 'string',
+                internals: password ? new Immutable.List([
+                    new Constraint.Enum([ password ])
+                ]) : def
+            }))
+            .set('host', new Parameter({
+                key: 'host',
+                type: 'string',
+                internals: host ? new Immutable.List([
+                    new Constraint.Enum([ host ])
+                ]) : def
+            }))
+            .set('hostname', new Parameter({
+                key: 'hostname',
+                type: 'string',
+                internals: hostname ? new Immutable.List([
+                    new Constraint.Enum([ hostname ])
+                ]) : def
+            }))
+            .set('port', new Parameter({
+                key: 'port',
+                type: 'string',
+                internals: port ? new Immutable.List([
+                    new Constraint.Enum([ port ])
+                ]) : def
+            }))
+            .set('pathname', new Parameter({
+                key: 'pathname',
+                type: 'string',
+                internals: pathname ? new Immutable.List([
+                    new Constraint.Enum([ pathname ])
+                ]) : def
+            }))
+            .set('search', new Parameter({
+                key: 'search',
+                type: 'string',
+                internals: search ? new Immutable.List([
+                    new Constraint.Enum([ search ])
+                ]) : def
+            }))
+            .set('hash', new Parameter({
+                key: 'hash',
+                type: 'string',
+                internals: hash ? new Immutable.List([
+                    new Constraint.Enum([ hash ])
+                ]) : def
+            }))
+    }
+
+    _constructorFromObj(obj) {
+        if (typeof obj === 'undefined' || obj === null) {
+            return this
+        }
+
+        let protocol = this._formatParam('protocol', obj.protocol)
+        let username = this._formatParam('username', obj.username)
+        let password = this._formatParam('password', obj.password)
+        let host = this._formatParam('host', obj.host)
+        let hostname = this._formatParam('hostname', obj.hostname)
+        let port = this._formatParam('port', obj.port)
+        let pathname = this._formatParam('pathname', obj.pathname)
+        let search = this._formatParam('search', obj.search)
+        let hash = this._formatParam('hash', obj.hash)
+
+        return this
+            .set('protocol', protocol)
+            .set('username', username)
+            .set('password', password)
+            .set('host', host)
+            .set('hostname', hostname)
+            .set('port', port)
+            .set('pathname', pathname)
+            .set('search', search)
+            .set('hash', hash)
```

Comment:
I'm not sure that `set` behaves like you expect it to when you pass it a dictionary as a first argument. Maybe using `withMutations` instead ?

e.g.
```js
return this.withMutations(url => {
    url.set(...)
})
```